src/udev/udevd.c

```diff
@@ -1735,6 +1691,147 @@ int main(int argc, char *argv[]) {
+        if (r < 0)
+                return log_error_errno(r, "failed to close excess fds: %m");
+
+        r = reset_all_signal_handlers();
+        if (r < 0)
+                return log_error_errno(r, "failed to reset signal handlers: %m");
+
+        r = reset_signal_mask();
+        if (r < 0)
+                return log_error_errno(r, "failed to reset signal mask: %m");
+
+        /* we don't reset any env vars */
+
+        /* create a barrier to wait for the daemon to be ready */
+        r = barrier_create(&barrier);
+        if (r < 0)
+                return log_error_errno(r, "failed to create barrier objcet: %m");
+
+        pid = fork();
+        switch (pid) {
+        case -1:
+                return log_error_errno(errno, "failed to fork off child: %m");
+        case 0:
+                /* child */
+                setsid();
+
+                pid = fork();
+                switch (pid) {
+                case -1:
+                        return log_error_errno(errno, "failed to fork off grandchild: %m");
+                case 0:
+                        /* grandchild (daemon) */
+                        barrier_set_role(&barrier, BARRIER_CHILD);
+
+                        /* connect /dev/null to stdin, stdout, stderr */
+                        if (log_get_max_level() < LOG_DEBUG) {
+                                _cleanup_close_ int null_fd = -1;
+                                null_fd = open("/dev/null", O_RDWR);
+                                if (null_fd >= 0) {
+                                        dup2(null_fd, STDIN_FILENO);
+                                        dup2(null_fd, STDOUT_FILENO);
+                                        dup2(null_fd, STDERR_FILENO);
+                                }
+                        }
+
+                        (void) umask(0);
+
+                        r = chdir("/");
+                        if (r < 0) {
+                                r = log_error_errno(errno, "could not change dir to /: %m");
+                                goto exit;
+                        }
+
+                        /* we do not write out a pidfile and we do not drop privileges */
+
+                        /* protect main daemon against the OOM killer */
+                        write_string_file("/proc/self/oom_score_adj", "-1000");
+
+                        /* start the actual daemon */
+                        r = main_notify(&barrier);
+                        if (r < 0)
+                                goto exit;
+
+exit:
+                        log_close();
+                        _exit(r < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
+                default:
+                        /* child */
+                        log_close();
+                        _exit(EXIT_SUCCESS);
+                }
+                break;
+        default:
+                /* parent */
+                barrier_set_role(&barrier, BARRIER_PARENT);
+
+                r = barrier_wait_next(&barrier);
```

Comment:
This should be:
`r = barrier_place_and_sync(&barrier);`

I should really improve the documentation and drop all the `barrier_*_next()` helpers.. they should really not be used by anyone.